### PR TITLE
Update Operator README for release

### PR DESCRIPTION
**Pull request by ravindk89**
Updated README to integrate some of the material from https://github.com/minio/docs/tree/master/source/kubernetes .

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,65 +1,256 @@
...
+public cloud infrastructure (Hybrid Cloud). The MinIO API maintains compatibility
+with the S3-API to simplify migration of applications onto MinIO Object Storage.
 
-- Kubernetes >= v1.17.0.
-- Create PVs.
-- Install `kubectl minio` plugin using `kubectl krew install minio`.
+## Table of Contents
 
-## Operator Setup
+* [Architecture](https://github.com/minio/operator#Architecture)
```
Comment (ravindk89) [inline README.md:11]: Hm. These links are broken. Might need to make them explicit as `operator/tree/master/README.md#` and see if that works...

File: README.md [hunk 2]
```diff
@@ -1,65 +1,256 @@
 # MinIO Operator [![Docker Pulls](https://img.shields.io/docker/pulls/minio/k8s-operator.svg?maxAge=604800)](https://hub.docker.com/r/minio/k8s-operator)
 
-MinIO Operator brings native support for [MinIO](https://github.com/minio/minio), [Graphical Console](https://github.com/minio/console), and [Encryption](https://github.com/minio/kes) to Kubernetes. This document explains how to get started with MinIO Operator using `kubectl minio` plugin.
+<a href="https://min.io"> <img src=https://raw.githubusercontent.com/minio/minio/master/.github/logo.svg width="600px"/> </a>
 
-## Prerequisites
+MinIO provides Kubernetes-native high performance object storage for private and
```
Comment (ravindk89) [inline README.md:5]: Will need @abperiasamy to comment on whether this wording makes sense.

File: README.md [hunk 3]
```diff
@@ -1,65 +1,256 @@
...
+IMAGE
+
+MinIO provides multiple methods for accessing and managing the MinIO Tenant:
+
+## MinIO Console
+
+The MinIO Console provides a graphical user interface (GUI) for interacting with
+MinIO Tenants. 
+
+IMAGE
```
Comment (ravindk89) [inline README.md:31]: ToDo: Update section for when Console for Operator (Tenant Management) is ready to include something like:

"Kubernetes Administrators can create and manage MinIO Tenants entirely through the Console and without relying on command line interactions." or something.

File: README.md [hunk 4]
```diff
@@ -1,65 +1,256 @@
...
 
-MinIO Operator offers MinIO Tenant creation, management, upgrade, zone addition and more. Operator is meant to control and manage multiple MinIO Tenants.
+# Architecture
 
-To get started, initialize the MinIO Operator deployment. This is a _one time_ process.
+Each MinIO Tenant represents an independent MinIO Object Store service within
+the Kubernetes cluster. The following diagram describes the architecture of as
+MinIO Tenant deployed into Kubernetes:
+
+IMAGE
```
Comment (ravindk89) [inline README.md:22]: Will create diagram if we first decide this looks OK.

File: README.md [hunk 5]
```diff
@@ -1,65 +1,256 @@
...
+  generates a Persistent Volume Claim (`PVC`) for each volume and evenly distributes
+  volumes across each `minio` pod. The example above results in 4 volumes per `minio` pod.
 
+- The `--capacity` field indicates the total capacity of the cluster. MinIO determines the
+  amount of storage to request for each `pvc` by dividing the specified capacity by the
+  total number of volumes in the server. The example above results in 1Ti requested
+  capacity per volume.
+
+
+# Kubernetes Cluster Configuration
```
Comment (ravindk89) [inline README.md:198]: Might be a good place to also stick in something around networking, like

## Networking Recommendations

MinIO can make efficient use of storage and can completely saturate network interfaces.

For example, consider a Node with locally-attached NVME SSD storage. The SSD can exhibit a maximum READ/WRITE performance of 2Gbps. If the node has 8 such drives, MinIO may drive a theoretical total of 16Gbps of traffic and overwhelm a 10GbE NIC. 

To avoid network bottlenecks, consider calculating the total possible READ/WRITE throughput of a single node and compare it against the total available network bandwidth provided by the node's NIC *and* the network infrastructure between that NIC and Clients. You can safely assume MinIO can drive up to 85% of theoretical drive throughput. 

Spitballing heavily here.

File: README.md [hunk 6]
```diff
@@ -1,65 +1,256 @@
...
+
+Username: admin 
+Password: abcd1234-abcd-1234-abcd-1234abcd1234
+
+Web interface access: 
+	$ kubectl port-forward svc/minio-tenant1-console 9090:9443
+	Point browser to https://localhost:9090
+
+Object storage access: 
+	$ kubectl port-forward svc/minio-tenant1-hl 9000:9000
```
Comment (Alevsk) [inline README.md:133]:
```suggestion
	$ kubectl port-forward svc/minio 9000:9000
```
TLDR; xd - Here service name should be only `minio`


When operator deploys a new tenant, 2 services are created, "minio" clusterIP service (has an IP address assigned) and "minio-tenant1-hl" headless service (doesnt have an IP address assigned), ie:

```
kubectl get svc
NAME                                           TYPE        CLUSTER-IP      EXTERNAL-IP   PORT(S)             AGE
kubernetes                                     ClusterIP   10.96.0.1       <none>        443/TCP             19d
minio                                          ClusterIP   10.98.115.236   <none>        443/TCP             24m
minio-tenant1-console   ClusterIP   10.97.71.11     <none>        9090/TCP,9443/TCP   23m
minio-tenant1-hl        ClusterIP   None            <none>        9000/TCP            24m
```
It's better for applications to access directly the `minio` service (K8S will LB connections to all minio pods) instead of the headless service, otherwise they will need to implement LB logic in the client side https://medium.com/faun/kubernetes-headless-service-vs-clusterip-and-traffic-distribution-904b058f0dfd

File: README.md [hunk 7]
```diff
@@ -1,65 +1,256 @@
...
+
+- The `Web interface access` opens the MinIO Console interface for interacting
+  with and managing the MinIO Tenant.
+
+- The `Object storage access` provides both the classic MinIO web interface *and* general 
+  API access to the MinIO Object Store.
+
+Applications inside the Kubernetes cluster can access the MinIO Tenant through either the
+web interface service (`minio-tenant1-console`) *or* the object storage service
+(`minio-tenant1-hl`).
```
Comment (Alevsk) [inline README.md:148]:
```suggestion
(`miniol`). 
```

File: README.md [hunk 8]
```diff
@@ -1,65 +1,256 @@
...
 ```sh
-kubectl minio tenant create --name tenant1 --servers 4 --volumes 16 --capacity 16Ti
+MinIO Tenant 'minio-tenant1' created
+
+Username: admin 
+Password: abcd1234-abcd-1234-abcd-1234abcd1234
+
+Web interface access: 
+	$ kubectl port-forward svc/minio-tenant1-console 9090:9443
+	Point browser to https://localhost:9090
```
Comment (Alevsk) [inline README.md:130]:
```suggestion
	$ kubectl port-forward svc/minio-tenant1-console 9443:9443
	Point browser to https://localhost:9443
```

Here we can use the same `9443` port

**Review by Alevsk — CHANGES_REQUESTED** [review]

File: README.md (hunk 6 above)
Comment (ravindk89) [reply README.md:133]: @nitisht @Alevsk  this might require a change to the operator PR

This was based on the output from #369 [here](https://github.com/minio/operator/pull/369/files#diff-22fba376bed03b38062f292edd3f54d976d78364ab8a1f3688f5b830acdd0a51R174), so maybe `minioSVCName` needs to be modified?

File: README.md [hunk 9]
```diff
@@ -1,65 +1,256 @@
...
+```sh
+   kubectl krew update
+   kubectl krew install minio
+```
+
+Run the following command to initialize the Operator:
 
 ```sh
 kubectl minio init
+
```
Comment (nitisht) [inline README.md:80]: This is actually `kubectl minio operator init`, can you fix this as well pls @ravindk89

**Comment by ravindk89** [comment]
Will update this once https://github.com/minio/docs/pull/314 closes

**Comment by ravindk89** [comment]
@Alevsk @eco-minio @nitisht updated for latest kubectl minio fixes. Apologies for the force push, i didn't want to reflog my way into fixing my commit history.

File: README.md [hunk 10]
```diff
@@ -1,65 +1,283 @@
...
+cp /var/run/secrets/kubernetes.io/serviceaccount/ca.crt /usr/local/share/ca-certificates/
+update-ca-certificates
 ```
 
-## Post Tenant Creation
+For applications *external* to the Kubernetes cluster, you must configure 
+[Ingress](https://kubernetes.io/docs/concepts/services-networking/ingress/) or a
+[Load Balancer](https://kubernetes.io/docs/concepts/services-networking/service/#loadbalancer) to
+expose the MinIO Tenant services. The `kubectl port-forward` command temporarily forwards traffic
```
Comment (nitisht) [inline README.md:168]:
```suggestion
expose the MinIO Tenant services. Alternatively, you can use the `kubectl port-forward` command to temporarily forward traffic 
```

File: README.md [hunk 11]
```diff
@@ -1,65 +1,283 @@
...
+for each PVC MinIO requests.
+
+You can estimate the number of PVC by multiplying the number of nodes in the 
+Tenant by the number of drives per node. For example, a 4-node Tenant with
+4 drives per node requires 16 PVC and therefore 16 PV.
+
+MinIO *strongly recommends* using locally-attached storage for each PV for
+object storage performance. MinIO recommends the following CSI drivers for
+creating local PV:
+
```
Comment (nitisht) [inline README.md:263]:
```suggestion
MinIO *strongly recommends* recommends the following CSI drivers for creating local PV:
```

**Review by nitisht — APPROVED** [review]
LGTM overall, some suggestions

File: README.md (hunk 11 above)
Comment (ravindk89) [reply README.md:262]: Resolved in separate commit to crunch two lines together.

**Review by ravindk89 — COMMENTED** [review]
@Alevsk @eco-minio if there are no further comments, I'll merge this and get started on a simple diagram just to round this readme out.

File: README.md (hunk 10 above)
Comment (ravindk89) [reply README.md]: Resolved in separate commit.

File: README.md [hunk 12]
```diff
@@ -1,65 +1,283 @@
...
+* [Architecture](https://github.com/minio/operator/blob/master/README.md#Architecture)
+* [Create a MinIO Tenant](https://github.com/minio/operator/blob/master/README.md#Create-a-MinIO-Tenant)
+* [Expand a MinIO Tenant](https://github.com/minio/operator/blob/master/REAMDE.md#Expand-a-MinIO-Tenant)
+* [Kubernetes Cluster Configuration](https://github.com/minio/operator/blob/master/README.md#Kubernetes-Cluster-Configuration)
 
-MinIO Operator offers MinIO Tenant creation, management, upgrade, zone addition and more. Operator is meant to control and manage multiple MinIO Tenants.
+# Architecture
 
-To get started, initialize the MinIO Operator deployment. This is a _one time_ process.
+Each MinIO Tenant represents an independent MinIO Object Store service within
```
[the remark this answers is not in the document]
Comment (ravindk89) [reply README.md]: Done

File: README.md (hunk 6 above)
Comment (ravindk89) [reply README.md]: This should now be addressed.

File: README.md [hunk 13]
```diff
@@ -1,65 +1,283 @@
...
+MinIO provides multiple methods for accessing and managing the MinIO Tenant:
+
+## MinIO Console
+
+The MinIO Console provides a graphical user interface (GUI) for interacting with
+MinIO Tenants. 
+
+IMAGE
+
+MinIO Tenant administrators can perform a variety of tasks through the Console,
```
Comment (eco-minio) [inline README.md:33]: Did we define Tenant administrators somewhere?  If not this might cause confusion, can probably switch slightly to "Administrators of MinIO tenants"

File: README.md [hunk 14]
```diff
@@ -1,65 +1,283 @@
...
+- This procedure assumes the cluster contains a
+  [`StorageClass`](https://github.com/minio/operator/blob/master/README.md#default-storage-class)
+  for the MinIO Tenant Persistent Volumes  (`PV`). The `StorageClass`
+  *must* have `volumeBindingMode: WaitForFirstConsumer`
+
+## 1) Install the MinIO Operator
+
+### Install using `kubectl krew`
+
+Run the following command to install the MinIO Operator and Plugin using `krew`:
```
Comment (eco-minio) [inline README.md:70]: Do we assume krew is a given within an administrators environment?  If not we should probably link to download or docs for those who are unfamiliar

File: README.md [hunk 15]
```diff
@@ -1,65 +1,283 @@
...
+      --storageClassName local-storage        \
+```
 
-In below example, we ask MinIO Operator to create a Tenant yaml with 4 nodes, 16 volumes, and 16 Ti total raw capacity (4 volumes of 1 Ti per node). This means you need to have 4 PVs of 1Ti each per node, with a total of 4 nodes, before attempting to create the MinIO tenant.
+- The `minio-tenant-1` argument specifies the name of the MinIO Tenant. The MinIO
+  Operator uses this name as a prefix for certain resources in the Tenant.
 
-Since MinIO has built-in erasure-code support for high-availability and data protection, we recommend the following CSI drivers for high-performance and scalability:
+- The `--servers` field indicates the number of `minio` pods to deploy into the cluster.
+  The cluster *must* have at least one available Node per `minio` pod.
```
Comment (eco-minio) [inline README.md:104]: Do we need to be pedantic and add "worker node" here?

File: README.md [hunk 16]
```diff
@@ -1,65 +1,283 @@
...
+  The cluster *must* have at least one available Node per `minio` pod. 
 
-- [Local PV](https://kubernetes.io/docs/concepts/storage/volumes/#local)
-- [OpenEBS Local PV](https://docs.openebs.io/docs/next/localpv.html)
-- [vSAN Direct](https://blogs.vmware.com/virtualblocks/2020/09/15/whats-new-in-vsan-7-update-1/)
+- The `--volumes` field indicates the total number of volumes in the Tenant. MinIO
+  generates a Persistent Volume Claim (`PVC`) for each volume and evenly distributes
+  volumes across each `minio` pod. The example above results in 4 volumes per `minio` pod.
+
+  If the cluster does not have sufficient Persistent Volume (`PV`) to bind to the
```
Comment (eco-minio) [inline README.md:110]:
```suggestion
  If the cluster does not have sufficient Persistent Volumes (`PV`) to bind to the
```

File: README.md [hunk 17]
```diff
@@ -1,65 +1,283 @@
...
+  generated `PVC`, Tenant creation hangs until the required `PV` exist.
```
Comment (eco-minio) [inline README.md:111]:
```suggestion
  generated `PVC`, Tenant creation hangs until the required `PV` exists.
```

File: README.md [hunk 18]
```diff
@@ -1,65 +1,283 @@
...
+
+- The `--capacity` field indicates the total capacity of the cluster. MinIO determines the
+  amount of storage to request for each `pvc` by dividing the specified capacity by the
+  total number of volumes in the server. The example above results in 1Ti requested
+  capacity per volume.
+
+  If the cluster does not have sufficient Persistent Volume (`PV`) to satsify the
```
Comment (eco-minio) [inline README.md:118]:
```suggestion
  If the cluster does not have sufficient Persistent Volumes (`PV`) to satisfy the
```

File: README.md [hunk 19]
```diff
@@ -1,65 +1,283 @@
...
+  requested capacity, Tenant creation hangs until the required `PV` exist.
```
Comment (eco-minio) [inline README.md:119]:
```suggestion
  requested capacity, Tenant creation hangs until the required `PV` exists.
```

File: README.md [hunk 20]
```diff
@@ -1,65 +1,283 @@
...
+
+- The `--namespace` field indicates the namespace onto which MinIO deploys the Tenant. 
+  If omitted, MinIO uses the `Default` namespace.
+
+  MinIO supports *one* MinIO Tenant per namespace.
+
+- The `--storageClassName` field indicates which `StorageClass` to use when generating
```
Comment (eco-minio) [inline README.md:126]: Link to storage class docs here?

File: README.md [hunk 21]
```diff
@@ -1,65 +1,283 @@
...
+For applications *external* to the Kubernetes cluster, you must configure 
+[Ingress](https://kubernetes.io/docs/concepts/services-networking/ingress/) or a
+[Load Balancer](https://kubernetes.io/docs/concepts/services-networking/service/#loadbalancer) to
+expose the MinIO Tenant services. The `kubectl port-forward` command temporarily forwards traffic
+from the local host to the MinIO Tenant. 
+
+- The `minio` service provides access to MinIO Object Storage operations.
+
+- The `minio-tenant-1-console` service provides access to the MinIO Console. The 
+  MinIO Console supports administration of the MinIO Tenant.
```
Comment (eco-minio) [inline README.md:174]:
```suggestion
  MinIO Console supports GUI administration of the MinIO Tenant.
```

File: README.md [hunk 22]
```diff
@@ -1,65 +1,283 @@
...
+  the MinIO Tenant.
+
+- This procedure assumes the cluster contains a
+  [`StorageClass`](https://github.com/minio/operator#default-storage-class)
+  for the MinIO Tenant Persistent Volume Claims (`PVC`).
 
-### Expanding a Tenant
+The following `kubectl minio` command expands a MinIO Tenant with an additional
+4 nodes, 16 volumes, and added capacity of 16Ti. This configuration requires
+*at least* 16 [Persistent Volumes](https://github.com/minio/operator#Local-Persistent-Volumes).
```
Comment (eco-minio) [inline README.md:192]: Should this say 16 new?  Also, it is possible for readers to misconstrue the requirement of 16 as a hard limit in minio rather than just what is required for this example

**Review by eco-minio — CHANGES_REQUESTED** [review]
Few clarifications and nits

File: README.md (hunk 17 above)
Comment (ravindk89) [reply README.md:111]: So I *think* that here, if we define `PV` as "Persistent Volumes", then "exist" would be correct. But its so damn ugly, that I'm more tempted to re-write this as

"If the cluster does not have an unbound Persistent Volume (`PV`) for each generated `PVC`, Tenant creation hangs until the required number of `PV`s exist."

`PV`s reads so weird but acronyms have that problem anyways.

File: README.md (hunk 19 above)
Comment (ravindk89) [reply README.md:119]: Same as above. I'm actually not sure of the exact grammar rules here (how do you pluralize an acronym anyways). Will chew on this.

File: README.md (hunk 13 above)
Comment (ravindk89) [reply README.md:33]: Done

File: README.md (hunk 14 above)
Comment (ravindk89) [reply README.md:70]: Added as part of the pre-requisites.

File: README.md (hunk 15 above)
Comment (ravindk89) [reply README.md:104]: I think it's worth being that specific. The Kubernetes docs uses the two terms interchangeably, but might as well avoid that potential edge.

File: README.md (hunk 17 above)
Comment (ravindk89) [reply README.md:111]: Tweaked to avoid this grammatical oddity entirely.

File: README.md (hunk 19 above)
Comment (ravindk89) [reply README.md:119]: Tweaked as per above

File: README.md (hunk 20 above)
Comment (ravindk89) [reply README.md:126]: Done

File: README.md (hunk 22 above)
Comment (ravindk89) [reply README.md:192]: Made some tweaks to hopefully make it clear this is driven by the number of PVC

**Review by ravindk89 — COMMENTED** [review]
@eco-minio back to you with fixes

**Review by eco-minio — DISMISSED** [review]
LGTM

File: README.md [hunk 23]
```diff
@@ -1,65 +1,295 @@
...
 
-This will add 32 drives spread uniformly over 8 servers to the tenant `tenant1`, with additional capacity of 32Ti. Read more about [tenant expansion here](https://github.com/minio/operator/blob/master/docs/expansion.md).
+To specify the storage class, include the `--storageClassName` option to 
+`kubectl minio tenant create`.
+
+
+## Local Persistent Volumes
+
+MinIO automatically creates Persistent Volume Claims (PVC) as part of Tenant creation.
+Ensure the cluster at least one
```
Comment (eco-minio) [inline README.md:265]:
```suggestion
Ensure the cluster has at least one 
```

File: README.md [hunk 24]
```diff
@@ -1,65 +1,295 @@
...
+
+
+## Local Persistent Volumes
+
+MinIO automatically creates Persistent Volume Claims (PVC) as part of Tenant creation.
+Ensure the cluster at least one 
+[Persistent Volume](https://kubernetes.io/docs/concepts/storage/persistent-volumes/)
+for each PVC MinIO requests.
+
+You can estimate the number of PVC by multiplying the number of `minio` server pods  in the
```
Comment (eco-minio) [inline README.md:269]:
```suggestion
You can estimate the number of PVC by multiplying the number of `minio` server pods in the 
```

**Review by eco-minio — CHANGES_REQUESTED** [review]
Two quick fixes needed

**Comment by ravindk89** [comment]
@eco-minio done - I have not tried github's squash + merge but it should work w/o breaking the tree. The PR was rebased just a few hours ago.

**Review by eco-minio — APPROVED** [review]
LGTM

**Review by Alevsk — APPROVED** [review]
LGTM
